### PR TITLE
[BUU] Update product images

### LINKED ISSUE #11065: [BUU] See and update product images
## Context
This issue introduces the ability to view and edit images in the product list.

## Description

**- As an:** enterprise user 
**- On page:** `/admin/products`
**- I want to be able to:** see and edit images
**- So that:** I can show the right pictures of my produce to customers

## Acceptance Criteria & Tests

**Scenario 1: Default images in list**
Given that I am in `/admin/products`
And I have products without an image
When my catalogue has loaded
Then I see the default illustration for products without an image

**Scenario 2: Start adding image**
Given that I am hovering on a product 
When I select to edit its image
Then I see a section with the current image at its full size (this could be either the default or a previously uploaded one)

**Scenario 3: Browsing for image**
Given that I am seeing the section where to edit the product image
When I select to upload an image
Then I see the default system upload window
And I can browse in my machine to find the image I want

**Scenario 4: Uploading image**
Given that I am browsing in my machine to find the image I want
When I select the image
Then my image starts to upload to my catalogue
And the default system window disappears
And the section where to edit the product image disappears
And I see a loading indicator

**Scenario 5: Successful image update**
Given that I am uploading a new product image
When the image is successfully uploaded
Then I see a success message
And I see the new image in my product catalogue

**Scenario 6: Error**
Given that I am uploading a new product image
When the image is not successfully uploaded
Then I see an error message describing what the issue is

@jibees can you please let me know which errors might happen? I'll design based on actual scenarios

## Design specs
A prototype is available here (upload image for product with default illustration, then click again to see the new image) > https://www.figma.com/proto/v1zbrWDZSRd3Nqoe0SJ2Sm/Engineering-Delivery?page-id=401%3A4320&type=design&node-id=401-4321&viewport=1013%2C303%2C0.14&t=tiXXkkRZsKsVBWnk-1&scaling=min-zoom&mode=design

Figma screens are available here > https://www.figma.com/file/v1zbrWDZSRd3Nqoe0SJ2Sm/Engineering-Delivery?type=design&node-id=401%3A4320&mode=design&t=rR08tEUkunFiHPm9-1

**New components and styles**

| Component | Example |
| --- | --- |
| Image in table (40x40) | Screenshot 2023-07-14 at 14 29 50|
| Edit image button |Screenshot 2023-07-14 at 14 31 05|
|Overlay dialog with title| Screenshot 2023-07-14 at 14 31 21|

### PR DESCRIPTION
**Pull request by dacook**
#### What? Why?

- Part of #11065
This also includes some visual updates.

I'm taking baby steps with this one so I can get dev input on the approach:
1.  The modal and form for the chosen image are loaded on-demand, with a Reflex
2. It submits a standard form to the existing ImagesController, because Reflex doesn't support file uploads. 
   * Perhaps we'll submit this with AJAX later, instead of reloading the whole page. But we'll probably focus on other parts of the page first.
3. When updated, it redirects back to the product list (keeping any existing query params)

#### What should we test?

As per https://github.com/openfoodfoundation/openfoodnetwork/issues/11065
But you can only update existing images, not add a new one yet.

#### Release notes

Because `admin_style_v3` has now been released to instance managers:

- [ ] User facing changes
- [ ] API changes (V0, V1, DFC or Webhook)
- [ ] Technical changes only
- [x] Feature toggled

The title of the pull request will be included in the release notes.

#### Dependencies
- #11942

### CODE REVIEW

**Comment by dacook** [comment]
Hi @mariocarabotta , this can be previewed on staging.

While working on this, I've generated some example errors by uploading a non-image file:

Screenshot 2024-01-11 at 3 26 45 pm

File: app/views/admin/products_v3/_edit_image.html.haml [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+  -# Submit to controller, because StimulusReflex doesn't support file uploads
+  = form_for [:admin, product, image], url: admin_product_image_path(product, image),
+      html: { multipart: true } do |f|
+    %input{ type: :hidden, name: :return_url, value: return_url}
+    = f.hidden_field :viewable_id, value: product.id
+
+    %p
+      = f.label :attachment, t(".upload"), class: "button primary icon-upload-alt"
+      = f.file_field :attachment, style: "display: none"
+      = f.submit # todo: submit automatically
```
Comment (mkllnk) [inline app/views/admin/products_v3/_edit_image.html.haml]: I think that the future will be to upload the files directly to the storage (AWS) from the browser. Rails has everything we need for that. The form submission then just contains the id of the uploaded file and a checksum and should work with StimulusReflex. The downside of this approach is probably that the storage validations gem doesn't work with this. It inspects the uploaded files and we would need to download it from the storage first. Still possible. We need to download it anyway to generate all the variants.

**Review by mkllnk — APPROVED** [review]
Nicely done.

**Comment by dacook** [comment]
@mkllnk , you pointed out that we can validate the image types on the client side. 
The browser can do this for us, so I suggest we add the simple mime-type condition `image/*`, and leave the more specific validation in the Rails model (rather than try to duplicate it).
https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/file#unique_file_type_specifiers

Sound ok to you?

**Comment by dacook** [comment]
Rebased to resolve conflict.

**Comment by mkllnk** [comment]
> The browser can do this for us, so I suggest we add the simple mime-type condition image/*, and leave the more specific validation in the Rails model (rather than try to duplicate it).

Yes, perfect.

**Review by mkllnk — APPROVED** [review]

**Comment by mariocarabotta** [comment]
> Hi @mariocarabotta , this can be previewed on staging.
> 
> While working on this, I've generated some example errors by uploading a non-image file:
> 
> Screenshot 2024-01-11 at 3 26 45 pm

is that actually one error with 2 different explanations?

File: app/reflexes/products_reflex.rb [hunk 2]
```diff
@@ -232,4 +241,10 @@ def product_finder(id)
   def current_id_from_element(element)
     element.dataset.current_id
   end
+
+  def image_params
```
Comment (rioug) [inline app/reflexes/products_reflex.rb:245]: Is this needed ? It is not used anywhere.

**Review by rioug — APPROVED** [review]
Looks good ! It just looks like there is a leftover method in the `products_reflex`

File: app/reflexes/products_reflex.rb (hunk 2 above)
Comment (dacook) [reply app/reflexes/products_reflex.rb:245]: 🙈 Oops, good pickup thanks!

**Comment by dacook** [comment]
> is that actually one error with 2 different explanations?

Yes, you've guessed correctly. I took a look and found that the second error has been added as a more user-friendly message. It looks like it was intended to replace the first one, but for some reason they both show. 
We can try to fix it when we work on error messages. I'll add a suggestion for this on the issue.

**Comment by dacook** [comment]
Awaiting dependency merged. before this can proceed.

**Comment by dacook** [comment]
Rebased, and added an update for hover behaviour and style fix.

**Pull request merged**